gap/prop.gi

```diff
@@ -38,58 +38,92 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
 InstallMethod(IsBridgelessDigraph, "for a digraph", [IsDigraph],
 D -> IsConnectedDigraph(D) and IsEmpty(Bridges(D)));
 
-InstallMethod(DIGRAPHS_IsMeetJoinSemilatticeDigraph, "for a homogeneous list",
-[IsHomogeneousList],
-function(nbs)
-  local i, j, k, n, x, len;
-
-  # The IsHomogenousList assumes the list <nbs> contains no or only empty lists
-  n := Length(nbs);
-  for i in [1 .. n] do
-    for j in [i + 1 .. n] do
-      if j in nbs[i] or i in nbs[j] then
-        continue;
-      fi;
-      x := Intersection(nbs[i], nbs[j]);
-      if IsEmpty(x) then
-        return false;
-      fi;
-      len := Length(x);
-      k := x[len];  # Check whether <k> is the meet of <i> and <j>
-      if Length(nbs[k]) < len then
+InstallMethod(DIGRAPHS_IsMeetJoinSemilatticeDigraph, "for a digraph and a bool",
+[IsDigraph, IsBool],
+function(D, join)
+  local P, U, ord, tab, S, N, i, x, T, l, q, z, y, nbs;
+
+  if not IsPartialOrderDigraph(D) then
+    return false;
+  fi;
+
+  D   := DigraphMutableCopyIfMutable(D);
+  N   := DigraphNrVertices(D);
+  tab := List([1 .. N], x -> []);  # table of meets/joins
+
+  if join then
+    nbs := OutNeighbours(D);
+    P   := DigraphTopologicalSort(D);
+    U   := OutNeighbours(DigraphReflexiveTransitiveReduction(D));
+  else
+    nbs := InNeighbours(D);
+    P   := Reversed(DigraphTopologicalSort(D));
+    U   := InNeighbours(DigraphReflexiveTransitiveReduction(D));
+  fi;
+
+  ord := [];
+  for i in [1 .. N] do
+    ord[P[i]] := i;
+  od;
+
+  S := [];
+
+  for x in P do
+    tab[x, x] := x;
+    for y in S do
+      T := [];  # EmptyPlist didn't improve speed
```

Comment:
I meant to remove the comments here from `prop.gi` and add them as GitHub comments. I will remove them later.